source/libtomo/accel/gpu/utils.cu

```diff
@@ -176,8 +183,32 @@ cuda_rotate_kernel(int32_t* dst, const int32_t* src, const float theta_rad,
                    const float theta_deg, const int nx, const int ny,
                    int eInterp = GPU_NN, cudaStream_t stream = 0)
 {
-    stream_sync(stream);
-    nppSetStream(stream);
+    // use static + comma operator to set the stream the first time
+    static bool _first = (nppSetStream(stream), false);
+    (void) _first;
+
+    bool _decrement = false;
+    // if stream is current stream continue
+    while(nppGetStream() != stream ||
+          _npp_stream_sync.load(std::memory_order_relaxed) == 0)
+    {
+        // when this hits zero we update stream
+        if(++_npp_stream_sync == 1)
+        {
+            _decrement = true;
+            nppSetStream(stream);
+            break;
+        }
+        else
+        {
+            --_npp_stream_sync;
+        }
+    }
+
+    if(nppGetStream() != stream)
+        throw std::runtime_error("Error! Wrong stream!");
+    // stream_sync(stream);
+    // nppSetStream(stream);
```

Comment:
```suggestion
```

Reply:
This was actually a bottleneck. The revised numbers that I noted (which reduced the recon time from ~570 seconds to 177 seconds) were solely a by-product on the scheme I implemented here with the `_npp_stream_sync`. This new scheme switches the stream in NPP less often and ensures that another thread does not swap out the NPP stream in the middle of this algorithm, which is why I was able to remove so many synchronizations.